src/sage/rings/polynomial/polydict.pyx

```diff
@@ -1467,68 +1726,75 @@ cdef class ETuple:
 
             sage: from sage.rings.polynomial.polydict import ETuple
             sage: e = ETuple([1,1,0])
-            sage: bool(e == loads(dumps(e)))
+            sage: e == loads(dumps(e))
             True
         """
         cdef size_t ind
-        # this is not particularly fast, but I doubt many people care
-        # if you do, feel free to tweak!
-        d = {self._data[2*ind]: self._data[2*ind+1]
-             for ind from 0 <= ind < self._nonzero}
-        return make_ETuple, (d, int(self._length))
+        d = {self._data[2 * ind]: self._data[2 * ind + 1] for ind in range(self._nonzero)}
+        return ETuple, (d, int(self._length))
 
     # additional methods
 
-    cpdef size_t unweighted_degree(self):
+    cpdef int unweighted_degree(self) except *:
         r"""
         Return the sum of entries.
 
-        ASSUMPTION:
-
-        All entries are non-negative.
-
         EXAMPLES::
 
              sage: from sage.rings.polynomial.polydict import ETuple
-             sage: e = ETuple([1,1,0,2,0])
-             sage: e.unweighted_degree()
+             sage: ETuple([1, 1, 0, 2, 0]).unweighted_degree()
              4
+             sage: ETuple([-1, 1]).unweighted_degree()
+             0
         """
-        cdef size_t degree = 0
+        cdef int degree = 0
         cdef size_t i
-        for i in range(1, 2*self._nonzero, 2):
-            degree += self._data[i]
+        for i in range(self._nonzero):
+            degree += self._data[2 * i + 1]
         return degree
 
     @cython.boundscheck(False)
     @cython.wraparound(False)
-    cdef size_t weighted_degree(self, tuple w):
+    cpdef int weighted_degree(self, tuple w) except *:
         r"""
         Return the weighted sum of entries.
 
         INPUT:
 
         - ``w`` -- tuple of non-negative integers
 
-        ASSUMPTIONS:
+        EXAMPLES::
+
+             sage: from sage.rings.polynomial.polydict import ETuple
+             sage: e = ETuple([1, 1, 0, 2, 0])
+             sage: e.weighted_degree((1, 2, 3, 4, 5))
+             11
+             sage: ETuple([-1, 1]).weighted_degree((1, 2))
+             1
 
-        ``w`` has the same length as ``self``, and the entries of ``self``
-        and ``w`` are non-negative.
+             sage: ETuple([1, 0]).weighted_degree((1, 2, 3))
+             Traceback (most recent call last):
+             ...
+             ValueError: w must be of the same length as the ETuple
         """
+        if len(w) != self._length:
+            raise ValueError('w must be of the same length as the ETuple')
+
         cdef size_t i
-        cdef size_t deg = 0
-        assert len(w) == self._length
-        for i in range(0, 2*self._nonzero, 2):
-            deg += <size_t> self._data[i+1] * <size_t> w[self._data[i]]
+        cdef int deg = 0
+        if len(w) != self._length:
+            raise ValueError
+        for i in range(0, 2 * self._nonzero, 2):
+            deg += self._data[i+1] * <int> w[self._data[i]]
```

Comment:
Essentially the same comment as for the other `range` just above.

Reply:
Indeed! This is **not** optimized!